src/object/srv_obj_migrate.c

```diff
@@ -258,18 +258,41 @@ migrate_pool_tls_destroy(struct migrate_pool_tls *tls)
 	if (tls->mpt_pool)
 		ds_pool_child_put(tls->mpt_pool);
 
-	d_rank_list_free(&tls->mpt_svc_list);
+	if (tls->mpt_svc_list.rl_ranks)
+		D_FREE(tls->mpt_svc_list.rl_ranks);
 
-	if (tls->mpt_clear_conts) {
+	if (tls->mpt_clear_conts)
 		d_hash_table_destroy_inplace(&tls->mpt_cont_dest_tab,
 					     true /* force */);
-	}
-
-	obj_tree_destroy(tls->mpt_root_hdl);
+	if (tls->mpt_done_cond)
+		ABT_cond_free(&tls->mpt_done_cond);
+	if (!daos_handle_is_inval(tls->mpt_root_hdl))
+		obj_tree_destroy(tls->mpt_root_hdl);
 	d_list_del(&tls->mpt_list);
 	D_FREE(tls);
 }
 
+void
+migrate_pool_tls_get(struct migrate_pool_tls *tls)
+{
+	tls->mpt_refcount++;
+}
+
+void
+migrate_pool_tls_put(struct migrate_pool_tls *tls)
+{
+	tls->mpt_refcount--;
+	if (tls->mpt_fini && tls->mpt_refcount == 1) {
+		if (tls->mpt_pool && tls->mpt_done_cond) {
+			ABT_mutex_lock(tls->mpt_pool->spc_lock);
+			ABT_cond_signal(tls->mpt_done_cond);
```

Comment:
Use eventual so that lock can be removed.

Reply:
Ah, yes. This does not need cross xstream to wait. I was dumb.  Thanks.